contracts/Proxy.sol

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2020 Compound Labs, Inc.
+
+pragma solidity ^0.7.5;
+
+import "./ErrorReporter.sol";
+import "./ProxyAdminStorage.sol";
+
+/**
+ * @dev This abstract contract provides a fallback function that delegates all calls to another
+ * contract using the EVM instruction `delegatecall`. We refer to the second contract as the
+ * _implementation_ behind the proxy.
+ *
+ * Additionally, delegation to the implementation can be triggered manually through the {_fallback}
+ * function, or to a different contract through the {_delegate} function.
+ *
+ * The success and return data of the delegated call will be returned back to the caller of the
+ * proxy.
+ */
+contract Proxy is ProxyAdminStorage, ErrorReporter {
+    /**
+     * @notice Emitted when pendingImplementation is changed
+     */
+    event NewPendingImplementation(
+        address oldPendingImplementation,
+        address newPendingImplementation
+    );
+
+    /**
+     * @notice Emitted when pendingImplementation is accepted, which means comptroller
+     * implementation is updated
+     */
+    event NewImplementation(address oldImplementation, address newImplementation);
+
+    /**
+     * @notice Emitted when pendingAdmin is changed
+     */
+    event NewPendingAdmin(address oldPendingAdmin, address newPendingAdmin);
+
+    /**
+     * @notice Emitted when pendingAdmin is accepted, which means admin is updated
+     */
+    event NewAdmin(address oldAdmin, address newAdmin);
+
+    constructor() {
+        // Set admin to caller
+        admin = msg.sender;
+    }
+
+    /*** Admin Functions ***/
+    function _setPendingImplementation(address newPendingImplementation) public returns (uint256) {
+        if (msg.sender != admin) {
+            return fail(Error.UNAUTHORIZED, FailureInfo.SET_PENDING_IMPLEMENTATION_OWNER_CHECK);
+        }
+
+        address oldPendingImplementation = pendingImplementation;
+
+        pendingImplementation = newPendingImplementation;
+
+        emit NewPendingImplementation(oldPendingImplementation, pendingImplementation);
+
+        return uint256(Error.NO_ERROR);
+    }
+
+    /**
+     * @notice Accepts new implementation. `msg.sender` must be pendingImplementation
+     * @dev Admin function for new implementation to accept its role as implementation
+     * @return uint 0=success, otherwise a failure (see ErrorReporter.sol for details)
+     */
+    function _acceptImplementation() public returns (uint256) {
+        // Check caller is pendingImplementation and pendingImplementation ≠ address(0)
+        if (msg.sender != pendingImplementation || pendingImplementation == address(0)) {
```

Comment:
Can msg.sender be ever 0?

Reply:
It can't, no. But this doesn't check that, it checks if any of these states is true and exits:

* the method was called from the wrong address
* the pending implementation hasn't been set yet